Use deps for setup ordering

JuantAldea opened the pull request:
# What does this implement/fix?

Topologically order codegen by dependencies

- Codegen
  - Schedule component to_code calls deterministically via a name-level
    dependency graph (Kahn’s algorithm), with logger-first and lexical
    tie-breaks.
  - Add platform-aware edges (platform component depends on target
    platform; target platform depends on esphome) when applicable.
  - Move #include "esphome.h" into core/config.py global phase; remove
    from writer.py to centralize include handling.
  - Emit USE_REGISTRATION_ORDER_SETUP to align runtime with codegen order.

- Runtime 
  - application.cpp: respect USE_REGISTRATION_ORDER_SETUP to skip priority
    sorting and keep registration order; add detailed DEBUG logs for setup
    order and per-component setup progress.
  - component.h: deprecate get_setup_priority/get_actual_setup_priority.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Relates to: https://github.com/esphome/backlog/issues/56 

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

JuantAldea committed:
Topologically order codegen by dependencies

- Codegen
  - Schedule component to_code calls deterministically via dependency graph
    with lexical tie-breaks
    - Schedule ESPHome core after the target-platform component
    - Schedule fixed-components before other components
  - Move #include "esphome.h" into core/config.py global phase; remove
    from writer.py to centralize include handling.

- Runtime (C++)
  - application.cpp: add detailed DEBUG logs for setup order and per-component
    setup progress.
  - component.h: deprecate get_setup_priority/get_actual_setup_priority.

JuantAldea commented:
Hello guys, this is my first time contributing to the project and I have put together some ideas for the topological sorting of dependencies for component initialization. I have also had some little fight getting the tests to work locally, so I have a added a separate commit for them.
I'm still quite unfamiliar with your test suite and I'm not sure what's the proper way of testing this change. All CI tests have passed on my own repository , and locally I have run:
-  script/test_build_components: passed
- script/unit: passed
- script/component : passed
- script/integration: some seem to fail randomly

I'd like to ask you how do you usually test these kind of changes.
This is still indeed a WIP as, at least, I have not yet fully removed the priority-based C++ code.

Any comments would be appreciated,
Cheers!

JuantAldea commented on script/test_build_components.py:
```diff
@@ -336,6 +369,23 @@ def run_grouped_test(
 
     try:
         result = subprocess.run(cmd, check=False)
+        print(f"Result code: {result.returncode}")
```
This file is full of bare prints.

JuantAldea commented on esphome/__main__.py:
```diff
@@ -413,7 +508,7 @@ def compile_program(args: ArgsProtocol, config: ConfigType) -> int:
 
 
 def upload_using_esptool(
-    config: ConfigType, port: str, file: str, speed: int
+    config: ConfigType, port: str, file: str | None, speed: int
 ) -> str | int:
```
The None case is handled a few lines below.

JuantAldea commented on script/test_build_components.py:
```diff
@@ -41,12 +41,21 @@
 
+# "esp32-s3-idf": 40,
+# "nrf52-adafruit": 40,
+# "esp32-c6-idf": 40,
```
Outdated.

jesserockz commented on esphome/core/config.py:
```diff
@@ -444,6 +444,7 @@ async def _add_platform_defines() -> None:
 
 @coroutine_with_priority(CoroPriority.CORE)
 async def to_code(config: ConfigType) -> None:
+    cg.add_global(cg.RawStatement('#include "esphome.h"'))
```
This header file is not for use within ESPHome itself. It is provided for lambdas or custom headers only. Including this file globally hides include issues inside other files

JuantAldea replied:
Got it. Then maybe I got it wrong, but from what I saw, the file is being unconditionally included at

https://github.com/esphome/esphome/pull/11266/files#diff-d6735e89ef07ce87e867a9149f6747850b177c33f9fb1b2b20fe699d3fa69f02L284

JuantAldea replied:
What I mean is that, when generating the main.cpp file, the #include "esphome" statement is generated, at the top of the file, by [1] whereas "using namespace esphome;" is generated by [2], right after the line you have highlighted.

Without [1], [2] produces an error, because the namespace does not exists.

From a main.cpp
```
// Auto generated code by esphome
// ========== AUTO GENERATED INCLUDE BLOCK BEGIN ===========
#include "esphome.h"           <---- emitted from [1]
using namespace esphome;       <---- emitted from [2]
using std::isnan;
[...]
```

There's also writer.py::get_include_text(), [3], but I could not find any calls to it.

Are there cases where core/config.py::to_code() [2] writes "using namespace esphome" and [1]  does not add the include? 

[1] https://github.com/esphome/esphome/blob/fa3ec6f732c3b94d67feba51a820197752074a0b/esphome/writer.py#L284

[2] https://github.com/esphome/esphome/blob/fa3ec6f732c3b94d67feba51a820197752074a0b/esphome/core/config.py#L467

[3] https://github.com/esphome/esphome/blob/fa3ec6f732c3b94d67feba51a820197752074a0b/esphome/writer.py#L62

jesserockz reviewed (changes requested):
I started looking at this, but you have too many unrelated changes in this PR. Please clean it up before marking as ready to review.

JuantAldea committed:
tests(integration): add build-time dependency-resolution tests and topo-order assertions

Add comprehensive integration tests and fixtures to validate dependency handling
and deterministic initialization order during build.

New fixtures (YAML):
- tests/integration/fixtures/building_cyclic_dependency.yaml
- tests/integration/fixtures/building_missing_auto_load_component.yaml
- tests/integration/fixtures/building_missing_dependency_component.yaml
- tests/integration/fixtures/building_topological_component_initialization_order.yaml

New external components (for fixtures):
- building_cyclic_dependency: DEPENDENCIES = ["building_cyclic_dependency"] (self-cycle)
- building_missing_auto_load_component: AUTO_LOAD = ["absent_component_xyz"]
- building_missing_dependency_component: DEPENDENCIES = ["absent_component_xyz"]
- building_topological_component_ordering: minimal component for order checks

New tests:
- tests/integration/test_building_topological_ordering.py
  - test_building_missing_dependency_component: success when building fails because      dependecy absent_component_xyz is missing.
  - test_building_missing_auto_load_component: success when builiding fails because auto-loaded component absent_component_xyz is missing.
  - test_building_topological_component_initialization_order: asserts log-reported
    initialization order respects ordering constraints.
  - test_building_cyclic_dependency: success when building fails because a cyclic dependency is found.

JuantAldea committed:
tests(integration): add compile_only fixture to run `esphome compile` and capture output

Add a pytest fixture `compile_only` to run `esphome compile` for a given
config and capture stdout, stderr, and the return code for assertions.

Motivation:
Enable compile-only integration tests that assert on compiler diagnostics,
warnings... without flashing or running firmware.

Usage example:
  out = await compile_only(config_path)
  assert out.returncode == 0
  assert "Building" in out.stdout

JuantAldea committed:
add missing type hints to AUTO_LOAD()

bdraco commented on esphome/core/component.h:
```diff
@@ -95,11 +95,12 @@ class Component {
+  [[deprecated]] virtual float get_setup_priority() const;
 
-  void set_setup_priority(float priority);
+  [[deprecated]] float get_actual_setup_priority() const;
```
This could likely be removed since it should only be used in core

bdraco commented on esphome/core/component.h:
```diff
@@ -95,11 +95,12 @@ class Component {
-  void set_setup_priority(float priority);
+  [[deprecated]] float get_actual_setup_priority() const;
+
+  [[deprecated]] void set_setup_priority(float priority);
```
This could likely be removed as well since its only called during codegen.

bdraco replied:
It means
```
    if CONF_SETUP_PRIORITY in config:
        add(var.set_setup_priority(config[CONF_SETUP_PRIORITY]))
```

would also need to be removed from `register_component`

bdraco replied:
And the whole setup_priority_overrides vector can be removed as well

bdraco replied:
which would make `ComponentPriorityOverride` dead code as well

bdraco commented on esphome/core/component.h:
```diff
@@ -95,11 +95,12 @@ class Component {
-  virtual float get_setup_priority() const;
 
-  float get_actual_setup_priority() const;
+  [[deprecated]] virtual float get_setup_priority() const;
```
This one truly does need to stay as removing it would break external components
